ops/_tracing/api.py

```diff
@@ -0,0 +1,142 @@
+        charm: ops.CharmBase,
+        tracing_relation_name: str,
+        ca_relation_name: str | None = None,
+        ca_data: str | None = None,
+    ):
+        """FIXME docstring.
+
+        Include `ops[tracing]` in your dependencies.
+
+        Declare the relations that the charm supports:
+
+        ```yaml
+        requires:
+            charm-tracing:
+                interface: tracing
+                limit: 1
+            send-ca-cert:
+                interface: certificate_transfer
+                limit: 1
+        ```
+
+        Initialise the tracing thing (FIXME) with the names of these relations.
+
+        ```py
+        import ops.tracing
+
+        class SomeCharm(ops.CharmBase):
+            def __init__(self, framework: ops.Framework):
+                ...
+                self._tracing = ops.tracing.Tracing(
+                    tracing_relation_name="charm-tracing",
+                    ca_relation_name="send-ca-cart",
+                )
+        ```
+        """
+        super().__init__(charm, f'{tracing_relation_name}+{ca_relation_name}')
+        self.charm = charm
+        self.tracing_relation_name = tracing_relation_name
+        self.ca_relation_name = ca_relation_name
+        self.ca_data = ca_data
+
+        # FIXME: Pietro recommends inspecting charm meta to validate the interface name
+        assert self.charm.meta.relations[tracing_relation_name].interface_name == 'tracing'
+        self._tracing = TracingEndpointRequirer(
+            self.charm,
+            tracing_relation_name,
+            protocols=['otlp_http'],
+        )
+
+        # TODO: COS requirer
+
+        for event in (
+            # FIXME: validate this
+            # - the start event may happen after relation-joined?
+            # - the k8s container died and got restarted
+            # - the vm [smth] crashed and got restarted
+            self.charm.on.start,
+            # In case the previous charm version has a relation but didn't save the URL
+            self.charm.on.upgrade_charm,
+            self._tracing.on.endpoint_changed,
+            self._tracing.on.endpoint_removed,
+        ):
+            self.framework.observe(event, self._reconcile)
+
+        self._certificate_transfer = None
+        if ca_relation_name:
+            self._certificate_transfer = CertificateTransferRequires(charm, ca_relation_name)
+            assert (
+                self.charm.meta.relations[ca_relation_name].interface_name
+                == 'certificate_transfer'
+            )
+            for event in (
+                self._certificate_transfer.on.certificate_set_updated,
+                self._certificate_transfer.on.certificates_removed,
+            ):
+                self.framework.observe(event, self._reconcile)
+
+    def _reconcile(self, _event: Any):
+        url, ca = self._get_config()
+        self.charm.set_tracing_destination(url=url, ca=ca)
+
+    def _get_config(self) -> tuple[str | None, str | None]:
```

Comment:
I'm not a huge fan of returning arbitrary pairs. Could we make this a namedtuple at least?

Reply:
yeah I was planning too...
I guess a dataclass is pretty much always better if I an to type the fields?
https://stackoverflow.com/a/50038614/705086